### PR TITLE
SPDY rate limiting

### PR DESCRIPTION
**Pull request by jpuigm**
[SPDY protocol spec](http://www.chromium.org/spdy/spdy-protocol/spdy-protocol-draft3-1#TOC-2.6.3-RST_STREAM) defines stream rate limiting.

Once a SPDY connection reaches `max_concurrent_streams` in flight streams, subsequent concurrent requests should be throttled using `RST_STREAM` -> `REFUSED_STREAM` (status code: `3`).

### CODE REVIEW

**Comment by essen** [comment]
Hey, the parsing/building of frames is done in cowlib (cow_spdy). The functions should also operate rather on a map for the list of settings, instead of setting aside one or two possible options.

**Comment by jpuigm** [comment]
Are you planning to merge https://github.com/ninenines/cowlib/pull/4 any time soon? It's open open since Dec 2013, and it's exactly what you meant.

**Comment by essen** [comment]
As you can see from the PR I asked for the corresponding Cowboy code which never happened. If they did send a Cowboy PR it would most likely be merged already.

**Comment by jpuigm** [comment]
I noticed that. I will then go ahead and update this PR using https://github.com/ninenines/cowlib/pull/4 changes, if you're happy with that.

**Comment by essen** [comment]
Don't know if it works but it doesn't look too bad so go with it.

**Comment by jpuigm** [comment]
Note that despite of using `layerhq:spdy-settings` you can go ahead and use/merge the original branch by `Etsukata`.

File: Makefile [hunk 1]
```diff
@@ -11,7 +11,7 @@ PLT_APPS = crypto public_key ssl
 # Dependencies.
 
 DEPS = cowlib ranch
-dep_cowlib = git https://github.com/ninenines/cowlib 1.2.0
+dep_cowlib = git https://github.com/layerhq/cowlib spdy-settings
```
Comment (essen) [inline Makefile]: This change will need to be removed.

**Comment by jpuigm** [comment]
I've pointed it out to `cowlib` `1.3.0` since that's the release that supports `settings`. You happy with that?

**Comment by essen** [comment]
Nope. Either you remove or you update Cowlib as part of a separate commit that also updates rebar.config. :-)

**Comment by jpuigm** [comment]
My bad! Totally forgot that you still had `rebar.config`.

**Comment by essen** [comment]
Ahah. But I mean the commit has to have the Makefile and rebar.config changes in it, not keep them out of sync (trying not to do that too often :-).

**Comment by jpuigm** [comment]
Sorry for the confusion.

**Comment by jpuigm** [comment]
Any updates here?

**Comment by essen** [comment]
Hey, I will get around to it soon but I have a few more pressing matters to attend to first. Thanks for your patience.

**Comment by essen** [comment]
Feel free to prepare other PRs meanwhile, if some things can be submitted independently. :-)

**Comment by essen** [comment]
Sent too soon.

I am currently working on an HTTP/1.1 test suite for RFC7230. I plan to do the same kind of test suite for SPDY immediately after and use that to ensure the SPDY implementation follows the spec properly.

By the way does anything implement SPDY 4 or just 3 and 3.1 nowadays?

**Comment by jpuigm** [comment]
We've never heard of anything implementing SPDY 4. We implement SPDY 3.1.

**Comment by essen** [comment]
A good patch for that would be advertising spdy/3.1 in addition to spdy/3. I don't think Cowboy does that yet. :-)

**Comment by jpuigm** [comment]
You mean in the README.md file? Or as a description of a new release?

**Comment by essen** [comment]
In the NPN negotiation (start_spdy). Though I suppose we don't actually check for the selected protocol, I think there's a ticket for that.

**Comment by essen** [comment]
I have cherry picked the Cowlib update. For the rest, it works, but the formatting is all messed up and I am not sure about some things yet so I will get back to you. Also I want to improve the Cowlib SPDY interface before making too many changes in Cowboy/Gun, in particular all these booleans are a problem.

**Comment by essen** [comment]
Closing this because we dropped SPDY in favor of HTTP/2. I'll make sure to look back at the PR when implementing the same feature in HTTP/2. Thanks!

**Pull request closed without merging**